### PR TITLE
COS Synchronization is performed at the end of initial scan.

### PR DESCRIPTION
**Pull request by tzezula**
COS Synchronization in J2SEProject is delayed to the end of the initial scan.

### CODE REVIEW

File: java.preprocessorbridge/src/org/netbeans/modules/java/preprocessorbridge/spi/CompileOnSaveAction.java [hunk 1]
```diff
@@ -281,12 +284,36 @@ public static Context update(
...
+         * @param isAllFilesIndexing true for all files indexing
+         * @param cacheRoot the cache root
+         * @param updated the changed files
+         * @param deleted the deleted files
+         * @param firer the fire callback
+         * @return the {@link Context} for update operation
+         * @since 1.50
+         */
+        @NonNull
+        public static Context update(
```
Comment (JaroslavTulach) [inline java.preprocessorbridge/src/org/netbeans/modules/java/preprocessorbridge/spi/CompileOnSaveAction.java]: OK.

File: java.preprocessorbridge/src/org/netbeans/modules/java/preprocessorbridge/spi/CompileOnSaveAction.java [hunk 2]
```diff
@@ -378,6 +405,15 @@ public static URL getTargetURL(@NonNull URL srcRoot) {
 
             return result;
         }
+
+        /**
+         * Returns true if the action represents an all files indexing.
+         * @return true for all files indexing
+         * @since 1.50
+         */
+        public boolean isAllFilesIndexing() {
```
Comment (JaroslavTulach) [inline java.preprocessorbridge/src/org/netbeans/modules/java/preprocessorbridge/spi/CompileOnSaveAction.java]: OK.

**Comment by jglick** [comment]
So this change only applies to Ant-based projects? Does anyone still use those any more??

**Comment by tzezula** [comment]
@jglick : Yes, just Ant based projects.
We are using them as a wrapper for mx GraalVM projects.

File: java.j2seproject/src/org/netbeans/modules/java/j2seproject/J2SEActionProvider.java [hunk 3]
```diff
@@ -883,8 +898,43 @@ private static String relativize(
...
+                     addGuard = true;
+                    deferredGuard = 1;
+                }
+            }
+            if (addGuard) {
+                final JavaSource js = createSource();
+                synchronized (this) {
+                    if (deferredGuard == 1) {
+                        try {
+                            js.runWhenScanFinished((cc) -> drainDeferred(), true);
```
Comment (jlahoda) [inline java.j2seproject/src/org/netbeans/modules/java/j2seproject/J2SEActionProvider.java]: Should deferredGuard be set to 2 here?

File: java.j2seproject/src/org/netbeans/modules/java/j2seproject/J2SEActionProvider.java [hunk 4]
```diff
@@ -703,7 +713,12 @@ private Boolean performUpdate(@NonNull final Context ctx) {
                                                 ex.getMessage()
                                             });
                                     }
-                                });
+                                };
+                                if (ctx.isAllFilesIndexing()) {
```
Comment (jlahoda) [inline java.j2seproject/src/org/netbeans/modules/java/j2seproject/J2SEActionProvider.java]: Here I wonder: should the work be deferred always? For short scans, that should be only a small delay (right?) and for big scans, wouldn't it be preferable to do whatever is necessary after the scan is finished?

File: java.j2seproject/src/org/netbeans/modules/java/j2seproject/J2SEActionProvider.java (hunk 3 above)
Comment (tzezula) [reply java.j2seproject/src/org/netbeans/modules/java/j2seproject/J2SEActionProvider.java]: Fixed.

File: java.j2seproject/src/org/netbeans/modules/java/j2seproject/J2SEActionProvider.java (hunk 4 above)
Comment (tzezula) [reply java.j2seproject/src/org/netbeans/modules/java/j2seproject/J2SEActionProvider.java]: The idea was that the changes will be applied as soon as possible and parallel to other scans.
The class and resource copying blocks project actions like `run`.
But I can try.

Comment (tzezula) [reply java.j2seproject/src/org/netbeans/modules/java/j2seproject/J2SEActionProvider.java]: It's not so easy. I will try in separate PR.

**Pull request merged**